components/gitpod-protocol/src/util/tracing.ts

```diff
@@ -76,15 +81,85 @@ export namespace TraceContext {
             return;
         }
 
-        ctx.span.addTags({
+        addNestedTags(ctx, {
             rpc: {
                 system: "jsonrpc",
                 method,
                 jsonrpc: {
-                    parameters: args.slice(),
+                    // parameters are added below, because we don't want them be nested, so that we have the possiblity to redact them.
                 },
             },
         });
+
+        for (let i = 0; i < args.length; i++) {
+            ctx.span.setTag(`rpc.jsonrpc.parameters.${i}`, args[i]);
+        }
+    }
+
+    /**
+     * Offers a way to opt-out of tracing for a particular JsonRPC parameter. Required if it contains PII, for instance.
+     * @param ctx
+     * @param indexes
+     * @returns
+     */
+    export function redactJsonRPCParameters(ctx: TraceContext, indexes: number[]) {
+        if (!ctx.span) {
+            return;
+        }
+
+        for (const i of indexes) {
+            ctx.span.setTag(`rpc.jsonrpc.parameters.${i}`, "REDACTED");
+        }
+    }
+
+    /**
+     * Does what one would expect from `span.addTags`: Calls `span.addTag` for all keys in map, recursively for objects.
+     * Example:
+     * ```
+     * TraceContext.addNestedTags(ctx, {
+     *    rpc: {
+     *       system: "jsonrpc",
+     *       jsonrpc: {
+     *          version: "1.0",
+     *          method: "test",
+     *          parameters: ["abc", "def"],
+     *       },
+     *    },
+     * });
+     * ```
+     * gives
+     * rpc.system = "jsonrpc"
+     * rpc.jsonrpc.version = "1.0"
+     * rpc.jsonrpc.method = "test"
+     * rpc.jsonrpc.parameters.0 = "abc"
+     * rpc.jsonrpc.parameters.1 = "def"
+     * @param ctx
+     * @param keyValueMap
+     * @returns
+     */
+    export function addNestedTags(ctx: TraceContext, keyValueMap: { [key: string]: any }, _namespace?: string) {
+        if (!ctx.span) {
+            return;
+        }
+        const namespace = _namespace ? `${_namespace}.` : '';
+
+        for (const k of Object.keys(keyValueMap)) {
+            const v = keyValueMap[k];
+            if (typeof v === 'object') {
+                addNestedTags(ctx, v, `${namespace}${k}`);
+            } else {
+                ctx.span.setTag(`${namespace}${k}`, v);
+            }
+        }
+    }
+
+    export function setOWI(ctx: TraceContext, owi: LogContext) {
+        if (!ctx.span) {
+            return;
+        }
+        addNestedTags(ctx, {
+            context: owi,
```

Comment:
This is sth we definitely need to align with the Go components (not namespaced there!). I'm definitely in favor of having `owi` namespaced somehow because:
 - it eases browsing of the fields in honeycomb
 - more unlikely to be completely overriden by a random `setTag`

No hard feelings about the name, though.

Reply:
So the way I'd usually go would multiple levels of namespacing. E.g. `workspace.owner.instance` that way you have something quite discoverable in your tool - you start writing `workspace.` and you'll get all attributes related to workspace.

Reply:
> So the way I'd usually go would multiple levels of namespacing

I understand the general id. But our core data structure is very flat (there is not much beyond user -> workspace -> workspaceInstance). And we already use this shape (with namespace `context`) in logging, for instance, for years. Also, all Go components use this for tracing (albeit not namespace - yet).

Happy to duplicate this field, or have additional convetions, but this one just makes so much sense. :-)

Reply:
For now, go with whatever is most consistent with what you have already. We'll then together across all teams work towards standardising on namespaces, but right now only worry about namespaces when you're introducing new attributes. I'm writing up the initial document now so we'll have something to iterate on  ☺️